File: apps/android/app/src/main/java/com/exponential/app/domain/RunOutcome.kt
package com.exponential.app.domain

import com.exponential.app.data.api.SteerDevice
import com.exponential.app.data.db.CodingSessionEntity

// EXP-637: how a FINISHED coding session presents itself, and when it can be
// resumed.
//
// An agent closes its own run out through the `exponential_sessions_end` MCP
// tool: a one-paragraph summary plus an outcome. Every client renders the same
// four labels for it (byte-equal across web, desktop, iOS and Android) — an
// ended row that carries no outcome (killed, merged, swept, or simply older
// than EXP-637) just reads "Ended".

/** The presentable outcome of an ended run — the glyph/tint selector too. */
enum class RunOutcome {
    Done,
    Blocked,
    NoChanges,

    /** Ended with no agent close-out: killed, merged, swept, or pre-EXP-637. */
    Ended,
    ;

    /** The one label every client shows for this outcome. */
    val label: String
        get() = when (this) {
            Done -> "Done"
            Blocked -> "Blocked"
            NoChanges -> "No changes"
            Ended -> "Ended"
        }
}

/**
 * The wire `outcome` value → its presentation. Anything unrecognized (a newer
 * server's value) falls back to [RunOutcome.Ended] rather than rendering a raw
 * wire token.
 */
fun runOutcomeOf(outcome: String?): RunOutcome = when (outcome) {
    DomainContract.codingSessionOutcomeDone -> RunOutcome.Done
    DomainContract.codingSessionOutcomeBlocked -> RunOutcome.Blocked
    DomainContract.codingSessionOutcomeNoChanges -> RunOutcome.NoChanges
    else -> RunOutcome.Ended
}

/** Shorthand for the label alone — the byte-equal string the row renders. */
fun runOutcomeLabel(outcome: String?): String = runOutcomeOf(outcome).label

/**
 * Where a Resume would go: the machine that ran the session, resolved against
 * the live device rows.
 */
data class RunResumeTarget(
    val sessionId: String,
    val deviceId: String,
    val deviceLabel: String,
)

/**
 * Whether [session] can be resumed right now, and on which machine (EXP-637).
 *
 * Everything the server checks, checked here first so the affordance is simply
 * absent instead of failing after the tap: the run must be the caller's OWN (a
 * teammate's run is never resumable — EXP-312), it must have ENDED (a live run
 * is steered, not resumed), it must carry the `device_id` of the machine that
 * ran it (that machine still holds the worktree and the agent's transcript),
 * and that machine must be online AND advertise the `resume-run` cap. The
 * machine may be a teammate's SHARED server (EXP-432) — the run is still the
 * caller's; own rows just win when the same machine id appears twice.
 *
 * Mirrors iOS `RunResume.target(for:devices:currentUserId:)`.
 */
fun resumeTargetFor(
    session: CodingSessionEntity,
    devices: List<SteerDevice>,
    currentUserId: String?,
): RunResumeTarget? {
    if (currentUserId == null || session.userId != currentUserId) return null
    if (session.status != DomainContract.codingSessionStatusEnded) return null
    val deviceId = session.deviceId?.takeIf { it.isNotEmpty() } ?: return null
    val candidates = devices.filter { it.deviceId == deviceId }
    val device = candidates.firstOrNull { it.isMine } ?: candidates.firstOrNull() ?: return null
    if (!device.online || !device.canResumeRun) return null
    return RunResumeTarget(
        sessionId = session.id,
        deviceId = device.deviceId,
        // Never blank: an unlabelled machine still reads as something.
        deviceLabel = device.deviceLabel.ifBlank { device.deviceId },
    )
}

File: apps/ios/ExpCore/Sources/Domain/RunOutcome.swift
import Foundation

/// EXP-637: how an ENDED coding session reads once the agent reported its
/// close-out (`exponential_sessions_end` stamps `summary` + `outcome`; since
/// EXP-673 it ENDS only an automation-started run — a person-started one
/// reports first and ends later, with its tab).
///
/// The labels are byte-identical across web (`lib/coding-session-display.ts`
/// `sessionOutcomeLabel`), desktop (`ended_runs.rs`), Android
/// (`domain/RunOutcome.kt`) and here — a run reads the same wherever it is
/// listed. Glyph and tint are per-platform (iOS draws them in ExpUI's
/// `EndedRunRow`); only the words are the contract.
public enum RunOutcomePresentation {
    /// The label a run's outcome wears. An ended row with NO outcome simply
    /// says "Ended": every path except the agent's own close-out (kill switch,
    /// tab close, PR merge, sweep) leaves the column NULL.
    public static func label(_ outcome: String?) -> String {
        switch outcome {
        case DomainContract.codingSessionOutcomeDone: return "Done"
        case DomainContract.codingSessionOutcomeBlocked: return "Blocked"
        case DomainContract.codingSessionOutcomeNoChanges: return "No changes"
        default: return "Ended"
        }
    }

    /// Whether an ENDED row carries the agent's close-out — the rule the runs
    /// lists filter on. Those are the rows with a summary worth expanding.
    /// Keyed on `outcome` (its only writer is `exponential_sessions_end`),
    /// NOT on `ended_by` (EXP-673): a person-started run's end is the tab
    /// close or kill that came after its report. Mirrored on web
    /// (`use-agents-data.ts`) and Android (`recentRunRows`).
    public static func hasCloseOut(_ session: CodingSessionEntity) -> Bool {
        session.status == DomainContract.codingSessionStatusEnded
            && session.outcome != nil
    }
}

/// EXP-637: whether an ended run can be picked up again, and on which machine.
///
/// Resume relaunches the pinned agent in the run's OWN worktree, so it is
/// bound to the machine that ran it: `steer.startSession({ resumeSessionId })`
/// refuses a foreign session (owner-only), a still-live one, another machine,
/// and one whose device does not advertise `resume-run`. The clients mirror
/// those gates so the affordance never appears where the send would bounce.
public enum RunResume {
    /// The machine a Resume would go to, or nil when this run cannot be
    /// resumed from here. Prefers the caller's OWN devices row when a shared
    /// teammate row happens to carry the same machine id.
    public static func target(
        for session: CodingSessionEntity,
        devices: [SteerDevice],
        currentUserId: String?
    ) -> SteerDevice? {
        guard CodingSessionOwnership.isOwn(session, userId: currentUserId),
              session.status == DomainContract.codingSessionStatusEnded,
              let deviceId = session.deviceId, !deviceId.isEmpty
        else { return nil }
        let candidates = devices.filter { $0.deviceId == deviceId }
        let device = candidates.first(where: \.isMine) ?? candidates.first
        guard let device, device.isOnline, device.canResumeRun else { return nil }
        return device
    }
}
